Sources/Sentry/SentryTracer.m

```diff
@@ -98,29 +152,76 @@ - (void)setDataValue:(nullable id)value forKey:(NSString *)key
 
 - (void)finish
 {
-    [_rootSpan finish];
-    [self captureTransaction];
+    [self finishWithStatus:kSentrySpanStatusUndefined];
 }
 
 - (void)finishWithStatus:(SentrySpanStatus)status
 {
-    [_rootSpan finishWithStatus:status];
-    [self captureTransaction];
+    _shouldBeFinished = true;
+    _finishStatus = status;
+    [self canBeFinished];
 }
 
-- (void)captureTransaction
+- (NSObject *)childrenLock
+{
+    return _parentTracer == nil ? _childrenLock : [_parentTracer childrenLock];
+}
+
+- (BOOL)hasUnfinishedChildren
 {
-    NSArray *spans;
-    @synchronized(_spans) {
-        // Make a new array with all finished child spans because if any of the transactions
-        // children is not finished the transaction is ignored by Sentry.
-        spans = [_spans
-            filteredArrayUsingPredicate:[NSPredicate
-                                            predicateWithBlock:^BOOL(id<SentrySpan> _Nullable span,
-                                                NSDictionary<NSString *, id> *_Nullable bindings) {
-                                                return span.isFinished;
-                                            }]];
+    @synchronized([self childrenLock]) {
+        for (id<SentrySpan> span in _spans) {
+            if (![span isFinished])
+                return YES;
+        }
+        return NO;
+    }
+}
+
+- (void)canBeFinished
+{
+    if (_waitForChildren && (!_shouldBeFinished || [self hasUnfinishedChildren]))
+        return;
+
+    [_rootSpan finishWithStatus:_finishStatus];
+    if (_parentTracer == nil) {
+        [self captureTransaction];
+    } else {
+        [_parentTracer canBeFinished];
     }
+}
+
+- (NSArray *)children
+{
+    NSMutableArray *result = [[NSMutableArray alloc] init];
+    @synchronized([self childrenLock]) {
+        for (id<SentrySpan> child in _spans) {
+            [result addObject:child];
+            if ([child isKindOfClass:[SentryTracer class]]) {
+                SentryTracer *childTracer = child;
+                [result addObjectsFromArray:[childTracer children]];
+            }
+        }
+    }
+    return result;
+}
+
+- (NSArray *)spans
+{
+    return _spans;
+}
+
+- (void)captureTransaction
+{
+    if (_hub == nil)
+        return;
+
+    NSArray *spans = [self.children
```

Comment:
`h`: I think we should use `_spans` instead here. `SentryTransaction` takes care of serializing all the children anyway. If we do it like this, I think it's going to mess things up. I also might be wrong.

Reply:
Added to https://github.com/getsentry/sentry-cocoa/issues/1134.